R/unitConverter.R

```diff
@@ -0,0 +1,103 @@
+#' Convert to common units
+#'
+#' @description
+#'
+#' When multiple (observed and/or simulated) datasets are present in a data
+#' frame, they are likely to have different units. This function helps to
+#' convert them to a common unit specified by the user.
+#'
+#' This is especially helpful while plotting since the quantities from different
+#' datasets to be plotted on the x-and y-axis need to have same units to be
+#' meaningfully compared.
+#'
+#' @return A data frame with measurement columns transformed to have common units.
+#'
+#' @param data A data frame (or a tibble).
+#' @param xUnit,yUnit Target units for `xValues` and `yValues`, respectively. If
+#'   not specified (`NULL`), first of the existing units in the respective
+#'   columns (`xUnit` and `yUnit`) will be selected as the common unit. For
+#'   available dimensions and units, see `ospsuite::ospDimensions` and
+#'   `ospsuite::ospUnits`, respectively.
+#'
+#' @seealso toUnit
+#'
+#' @examples
+#'
+#' # small dataframe to illustrate the conversion
+#' (df <- dplyr::tibble(
+#'   xValues = c(15, 30, 60), xUnit = "min", xDimension = "Time",
+#'   yValues = c(0.25, 45, 78), yUnit = c("", "%", "%"), yDimension = c("Fraction", "Fraction", "Fraction"),
+#'   molWeight = c(12.5, 10, 5)
+#' ))
+#'
+#' ospsuite:::.unitConverter(df)
+#' ospsuite:::.unitConverter(df, xUnit = ospUnits$Time$h)
+#' ospsuite:::.unitConverter(df, yUnit = ospUnits$Fraction$`%`)
+#' ospsuite:::.unitConverter(df, xUnit = ospUnits$Time$h, yUnit = ospUnits$Fraction$`%`)
+#'
+#' @keywords internal
+#' @noRd
+.unitConverter <- function(data, xUnit = NULL, yUnit = NULL) {
+  # No validation of inputs for this non-exported function.
+  # All validation will take place in the `DataCombined` class itself.
+
+  # The observed and simulated data should have same units.
+  #
+  # Therefore, if target units are not specified, we need to choose one for
+  # consistency. For no special reason, first element from a vector of unique
+  # units will be selected.
+  targetXUnit <- xUnit %||% .selectDefaultTargetUnit(data$xUnit)
+  targetYUnit <- yUnit %||% .selectDefaultTargetUnit(data$yUnit)
+
+  # *WARNING*: Do not change the order of two `mutate()` statements.
+  #
+  # Since the old `xUnit` and `yUnit` columns are need for unit conversion,
+  # those columns must be updated only *after* the conversions have taken place.
+  data %>%
+    dplyr::rowwise() %>% # transform values to common units
+    dplyr::mutate(
+      xValues = toUnit(
```

Comment:
I am wondering.... Will it call `toUnit` for every single value? In that case it appears to me as super inefficient. Image a population simulan with thousands of values - will it make thousands of calls to `toUnit`? It would be much better to convert a whole array with only one call.

Reply:
> In that case it appears to me as super inefficient

This is the case, but the problem is that `toUnit` has only single parameter vectorized, but not the others, and this doesn't play well with data frame workflows.

I can get rid of this efficiency, but it's comes at the cost of increased code complexity.

Reply:
What other parameters should be vectorized?